### ISSUE TITLE
Consider adding a ToObject class

### ISSUE THREAD

**Issue by iand675** [bos/aeson#544]
I semi-regularly find myself using APIs that support attaching some form of user-provided metadata, but the metadata must be an object. As an example, the Segment HTTP API has a number of properties that just take grab-bags of user metadata, so my API is looking about like this:

``` haskell
track :: ToJSON a => SegmentClient -> Context -> EventName a -> a -> IO ()
track c ctxt n val = ... v
  where
    (Object v) = toJSON val
```

I'm not terribly crazy about having to do partial matching just to extract a value, and think it would be lovely if there was a class that be an instance only for data types that made sense to be generated as objects that could use the same TH or Generic instance generation:

```
class ToObject a where
  toObject :: a -> Object
```

**Comment by phadej** [comment]
I'm :+1:, I have some use cases where I build an objects. Also e.g. `log` package would benefit from it.

This is however quite big change if done properly (generic/TH, `toEncoding` needed changes etc.)

**Comment by iand675** [comment]
Certainly, it could be a big change. If this was implemented, it might make sense to provide similar for each of the other base JSON types as well. I've wished for this issue to be fixed multiple times now though, so seemed worth discussing.

**Comment by bergmark** [comment]
I need some time to think about this, so I'll follow up later on with questions. Do continue the discussion in the meantime :)

**Comment by phadej** [comment]
(I have also `ToJSONField` class which let do `Maybe` field omission more cleanly)

That said: i'm not 100% sure that needs to be in `aeson`, all could be built on top
